File: 3777-maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k/maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.py
class Solution:
    def maxProduct(self, nums: List[int], k: int, limit: int) -> int:
        ans = -1 
        n = len(nums)
        ppos = set()
        pneg = set()
        zpos = set()
        zneg = set()
        for i, x in enumerate(nums): 
            pos = set()
            neg = set()
            if x <= limit: pos.add((x, x))
            for s, p in ppos: 
                ss = s - x
                pp = p * x
                if pp <= limit: neg.add((ss, pp))
            for s, p in pneg: 
                ss = s + x
                pp = p * x
                if pp <= limit: pos.add((ss, pp))
            ppos |= pos
            pneg |= neg 
            pos = {(x, x == 0)}
            neg = set()
            for s, tf in zpos: 
                ss = s - x
                tt = tf or x == 0 
                neg.add((ss, tt))
            for s, tf in zneg: 
                ss = s + x
                tt = tf or x == 0
                pos.add((ss, tt))
            zpos |= pos
            zneg |= neg
        for s, p in ppos | pneg: 
            if s == k: ans = max(ans, p)
        for s, tf in zpos | zneg: 
            if s == k and tf: ans = max(ans, 0)
        return ans 

File: 3777-maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k/maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, unordered_map<int, unordered_map<int, int>>>>dp;

    int tdp(int in, int isodd, int as, int prod, vector<int>& A, int k, int lim){
        if(in==A.size()){
            return ((prod<=lim) && (as==k) && isodd)?prod:INT_MIN;
        }
        if(dp.count(in) && dp[in].count(isodd) && dp[in][isodd].count(as) && dp[in][isodd][as].count(prod))return dp[in][isodd][as][prod];

        int ans=tdp(in+1, isodd, as, prod, A, k, lim);
        if(isodd==0)ans=max(ans, tdp(in+1, 2, as+A[in], A[in], A, k, lim));
        else if(isodd==1)ans=max(ans, tdp(in+1, 2, as+A[in], min(A[in]*prod, lim+1), A, k, lim));
        else ans=max(ans, tdp(in+1, 1, as-A[in], min(A[in]*prod, lim+1), A, k, lim));
        return dp[in][isodd][as][prod]=ans;
    }

    int maxProduct(vector<int>& A, int k, int limit) {
        int ts=0;
        for(auto&a:A)ts+=a;
        if(ts<k || -ts>k)return -1;
        int ans = tdp(0, 0, 0, 0, A, k, limit);
        return (ans==INT_MIN)?-1:ans;
    }
};